apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1342,21 +1377,265 @@ private void StartPush()
+                    {
+                        // Update backoff of retryableErrors Ids in _retryableIds dictionary.
+                        UpdateBackoff(retryableErrors.Select(j => j.Id), backoff);
+                        // Retry errors after backOff.
+                        Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                        Add(delayTask, new NextAction(false, () =>
+                        {
+                            // This will add the extendIds to the corresponding queue and start push to retry those ids after delay.
+                            // We don't need recursion as the handling of ack response will take care of retry and _retryableIds
+                            // dictionary will ensure that we don't retry after 10 seconds.
+                            actionToRetry(retryableErrors);
+                            StartPush();
+                        }));
+                    }
+                }
+            }
+
+            private void UpdateBackoff(IEnumerable<string> retryableErrors, TimeSpan? backoff)
+            {
+                foreach (var item in retryableErrors)
+                {
+                    if (_retryableIds.TryGetValue(item, out var info))
+                    {
+                        // _retryableIds is a concurrent dictionary but we are updating the value.
+                        _retryableIds.Locked(() => _retryableIds.TryUpdate(item, new RetryInfo(info.FirstTimeOfFailureInUtc, backoff), info));
+                    }
+                }
+            }
+
+            private void HandleAckResponseForExactlyOnce(Task writeTask, List<string> ackIds, List<string> nackIds, List<TimedId> extendIds)
+            {
+                _concurrentPushCount -= 1;
+                _pushInFlight -= ackIds?.Count ?? 0 + nackIds?.Count ?? 0 + extendIds?.Count ?? 0;
+
+                bool hasAckIds = ackIds?.Count > 0;
+                bool hasNackIds = nackIds?.Count > 0;
+                bool hasExtendIds = extendIds?.Count > 0;
+                // In one call, we 'll have one of acks or nacks or extends, not all.
+                var ids = hasAckIds ? ackIds
+                    : hasNackIds ? nackIds : extendIds.Select(j => j.Id);
+
+                if (writeTask.IsFaulted)
+                {
+                    // There may be few successes and few failures or all failures.
+                    var rpcException = writeTask.Exception.As<RpcException>();
+                    if (rpcException != null)
+                    {
+                        var ackError = rpcException.GetAckError(ids);
+                        // Only temporary failures should be retried. Mark them as retryable if not already.
+                        UpdateRetryableIds(ids, ackError);
+
+                        if (hasAckIds)
+                        {
+                            RetryAndHandleAcks(true, ackError);
+                        }
+                        else if (hasNackIds)
```

Comment:
What happens if we have both acks and nacks? Do we not want to retry the nacks in that case?

Reply:
My understanding was that we should have one of acks or nacks or extends in one call. I have now refactored the method to mimic the existing HandleAckResponse method.